### DISCUSSION TITLE
ideas for templating hubverse packages

### DISCUSSION THREAD

**Discussion by nickreich** [hubverse-org/.github#4] [Ideas]
There have been a number of ideas for standardizing development workflows for packages we develop. Just trying to catalogue the ideas proposed here so we don't lose track of them. If I understand correctly, some of these things, e.g. the package styling, may require a separate package whose job is just this, where other items may be able to be included in some kind of package template repo. It's also not clear just now many new packages we may have, but likely at least a few given hubData and hubAdmin ideas... And some of the items in here could be merged in to existing packages as well.

- lintr template
- package styling (CSS, logos, etc...)
- other misc file templates. although many of these can be installed easily with `usethis`, having some standard set of steps of how things should be constructed and what tools to use could be useful, e.g.: 
    - pkgdown branch/actions
    - R CMD CHECK workflow and options
    - LICENSE choices
    - any specific .Rbuildignore and .gitignore things (e.g. *.DSStore and/or renv files)
    - Code of Conduct/Community Standards text
    - Contribution guidelines
- ... (maybe other things)?

**Comment by LucieContamin** [comment]
Should we also have some kind of README template? With some common information (logo, expected badges, links to HubDoc? or Code of conducts, etc.) 
And also a testthat/code coverage branch/actions?

**Reply by annakrystalli** [reply]
@LucieContamin See this template README for what I've implemented so far: https://github.com/Infectious-Disease-Modeling-Hubs/testPkg

**Comment by annakrystalli** [comment]
I've created the `hubDevs` package that does a lot of the above:  https://github.com/Infectious-Disease-Modeling-Hubs/hubDevs

I'm just adding some documentation to make it clearer what it does and will need to add to it once the package documentation styling package but much of the templating is done and many functions can be used separately to update existing files (linke `CODE_OF_CONDUCT.md` and `CONTRIBUTING.md`

**Comment by annakrystalli** [comment]
The hubDevs  README is now updated with demo of package functionality https://github.com/Infectious-Disease-Modeling-Hubs/hubDevs

Basically two function calls will get you the following skeleton package, locally and on GitHub in the hubverse org: https://github.com/Infectious-Disease-Modeling-Hubs/testPkg

**Comment by annakrystalli** [comment]
I've now created the `hubStyle` package too (https://github.com/Infectious-Disease-Modeling-Hubs/hubStyle) to create hubverse themed documentation sites and added experimental functionality to `hubDevs` to set it all up.

Would welcome specific feedback on the website template too.

Here's a preview of `hubStyle` in action: https://65ae9052bb7760200e3332f2--hubverse-pkgdown-preview.netlify.app/

**Reply by LucieContamin** [reply]
Thank you for the "style" package I like it! I just have some small questions but not sure you can do anything about it!
- on the header can we have more space on top of the package name? (on my computer it looks really close to the top bar of my browser)
- on the home page, for the logo, I wonder if we can lower the horizontal bar so that it does not go through the logo? does that make sense?

**Reply by nickreich** [reply]
- +1 to both of Lucie's suggestions.
- I would advocate for a less colorful banner at the top. While I like the blue shading, I also think that keeping it visually simpler is good. For example, that would be more consistent with the hubDocs pages as well as with other proejcts, e.g. tidyverse. 
- I like the small sizing and font better on the current Docs. I wonder if we could steal some of the CSS from over there and apply it here. That might also ensure a bit more consistency, even though for the moment we are using different systems to host these different kinds of documentation.

**Comment by annakrystalli** [comment]
In particular, the footer can be a good place to add repetitive but useful links to further info, including more info on funding, disclaimers, links to other useful docs etc.

Here as an example is what rOpenSci have on their footer: 
<img width="754" alt="image" src="https://github.com/Infectious-Disease-Modeling-Hubs/.github/assets/5583057/ef7d8759-7bab-4c1f-a993-71a3560912a4">

**Reply by LucieContamin** [reply]
I like the footer idea

**Reply by nickreich** [reply]
I do too!

**Comment by annakrystalli** [comment]
Thanks for the feedback @LucieContamin & @nickreich .

I've implemented feedback as best I can in updated hubStyle pkg which now looks like this: https://65aff60c680cda0fac829d5a--hubverse-pkgdown-preview.netlify.app/

Let me know what you think!

**Reply by nickreich** [reply]
I think it looks great! Should we merge for now and file an issue for later about adding a footer?

**Reply by LucieContamin** [reply]
I like it too, I just have a quick question, on the articles "sub-menu", the space in between section is a little bit big, is that normal?

**Reply by annakrystalli** [reply]
@nickreich Great! To be honest I've been working directly in the `main` branch for this one 😬. Not best practice I know but faster and thought it'd be ok as none of this work is critical.

Re: issue for further enhancements, yes! 

Overall my plan is to use the fact I have to examine each repo individually to upgrade it to the new organisation name, to standardise the doc sites, READMEs and community docs across all our packages at the same time. Hence I've put a bit of effort into an experimental `hubDevs` and `hubStyle` package the last few days.

The nice thing about this template approach though is that any future improvements we make will propagate automatically to all our doc sites the next time they are built! 🚀

@LucieContamin that is the standard sub menu spacing. It is a bit awkward indeed though. Feel free to file an issue but I likely won't get round to it till a bit later.

**Reply by nickreich** [reply]
Ha! I hadn't noticed, but that main branch work seems fine to me here. All sounds great thanks for all your work on this!

**Reply by nickreich** [reply]
@annakrystalli if you haven't already done it, I'll file an issue for the footer.
